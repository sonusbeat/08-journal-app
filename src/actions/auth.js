import types from "../types/types";
import firebase, { googleAuthProvider } from  "../firebase/firebase-config";

export function startLoginEmailPassword( email, password ) {

  return ( dispatch ) => {

    firebase.auth().signInWithEmailAndPassword( email, password )
      .then( ({ user }) => {
        dispatch( login(user.uid, user.displayName) );
      })
      .catch(error => console.log(error));

  };

}

export function startRegisterWithEmailPasswordName( email, password, name ) {
  return ( dispatch ) => {
    firebase.auth().createUserWithEmailAndPassword( email, password )
      .then( async({ user }) => {

        // Firebase Function to update Display Name
        await user.updateProfile({ displayName: name });

        dispatch( login(user.uid, user.displayName) );
      })
      .catch( error => console.log(error) );
  };
}

export function startGoogleLogin() {
  return ( dispatch ) => {
    firebase.auth().signInWithPopup( googleAuthProvider )
      .then( ({ user }) => {
        dispatch( login( user.uid, user.displayName ) );
      });
  };
}

export default function login( uid, displayName ) {
  return {
    type: types.login,
    payload: { uid, displayName }
  };
}