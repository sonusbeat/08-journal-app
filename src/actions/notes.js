import { db } from "../firebase/firebase-config";
import types from '../types/types';
import { loadNotes } from '../helpers/loadNotes';

export const startNewNote = () => {
  return async ( dispatch, getState ) => {

    const uid = getState().auth.uid;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    // Guardar la nota en Firestore
    const doc = await db.collection(`${ uid }/journal/notes`).add( newNote );

    // Disparar la acciÃ³n al reducer
    dispatch( activeNote ( doc.id, newNote ) );

  };
};

export const activeNote = ( id, note ) => {
  return {
    type: types.notesActive,
    payload: { id, ...note }
  };
};

export const startLoadingNotes = ( uid ) => {

  return async (dispatch) => {

    const notes = await loadNotes( uid );

    dispatch( setNotes( notes ) );

  };

};

export const setNotes = ( notes ) => ({
  type: types.notesLoad,
  payload: notes
});

export const startSaveNote = ( note ) => {
  return async ( dispatch, getState ) => {
    const { uid } = getState().auth;

    if ( !note.url ) delete note.url;

    const noteToFirestore = { ...note };

    // Eliminar el id para evitar problemas en firebase
    delete noteToFirestore.id;

    // Guardar en Firestore
    db.doc(`${uid}/journal/notes/${ note.id }`).update( noteToFirestore );
  };
};
