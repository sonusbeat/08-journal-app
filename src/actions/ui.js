import types from '../types/types';

export function setError( error ) {
  return {
    type: types.uiSetError,
    payload: error
  };
};

export function unsetError() {
  return {
    type: types.uiRemoveError
  };
};

export function startLoading() {
  return {
    type: types.uiStartLoading
  };
}

// Nueva Sintaxis ES6
export const finishLoading = () => ({
  type: types.uiFinishLoading
});