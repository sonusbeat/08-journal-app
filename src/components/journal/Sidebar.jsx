import { useDispatch, useSelector } from 'react-redux';

import JournalEntries from './JournalEntries';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

const Sidebar = () => {
  const dispatch = useDispatch();
  const { name } = useSelector( state => state.auth );

  const handleLogout = () => {
    dispatch( startLogout() );
  };

  const HandleAddNew = () => {
    dispatch( startNewNote() );
  };

  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">

        <h2 className="journal__sidebar-title">
          <i className="far fa-user mr-1"></i>
          <span>{ name }</span>
        </h2>

        <button
          id="logoutBtn"
          type="button"
          className="btn btn-primary"
          onClick={ handleLogout }
        >Logout</button>

      </div>

      <div
        id="newEntryBtn"
        className="journal__sidebar-new-entry"
        onClick={ HandleAddNew }
      >
        <i className="far fa-calendar-plus fa-5x mb-2"></i>
        <p>New Entry</p>
      </div>

      <JournalEntries />
    </aside>
  );
}

export default Sidebar;