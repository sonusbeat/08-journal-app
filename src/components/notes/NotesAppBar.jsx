import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector( state => state.notes );

  const handleSave = () => {
    dispatch( startSaveNote( note ) );
  };

  const handlePictureUpload = () => {
    document.getElementById("fileSelector").click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if ( file ) {
      dispatch( startUploading( file ) );
    }
  };

  return (
    <div className="notes__app-bar">
      <span>15 de Abril 2017</span>

      <input
        id="fileSelector"
        name="image"
        type="file"
        style={{ display: "none" }}
        onChange={ handleFileChange }
      />

      <div>
        <button
          className="btn"
          onClick={ handlePictureUpload }
        >
          Picture
        </button>

        <button
          className="btn"
          onClick={ handleSave }
        >
          Save
        </button>
      </div>
    </div>
  );
}

export default NotesAppBar;