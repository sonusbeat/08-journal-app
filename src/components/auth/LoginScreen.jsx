import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import useForm from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { setError, unsetError } from '../../actions/ui';
import validator from "validator";

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { msgError, loading } = useSelector( state => state.ui );

  const [ formValues, handleInputChange ] = useForm({
    email: "",
    password: "",
  });

  const { email, password } = formValues;

  const isFormValid = () => {
    if ( validator.isEmpty( email ) ) {
      dispatch( setError("Email should be present !") );
      return false;
    } else  if ( !validator.isEmail( email ) ) {
      dispatch( setError("Email format not valid !") );
      return false;
    } else  if ( validator.isEmpty( password ) ) {
      dispatch( setError("Password should be present !") );
      return false;
    } else  if ( password.trim().length < 8 ) {
      dispatch( setError("Password should be greater than 8 characters !") );
      return false;
    }

    dispatch( unsetError() );

    return true;
  };

  const handleLogin = (e) => {
    e.preventDefault();

    if( isFormValid() ) {
      dispatch( startLoginEmailPassword( email, password ) );
    }
  };

  const handleGoogleLogin = () => {
    dispatch( startGoogleLogin() );
  };


  return (
    <>
      {
        msgError &&
        (
          <div className="auth__alert-error mb-2 text-center">
            { msgError }
          </div>
        )
      }

      <h3 className="auth__title">Login</h3>

      <form onSubmit={ handleLogin }>

        <input
          name="email"
          value={ email }
          className="auth__input"
          type="text"
          placeholder="email"
          autoComplete="off"
          autofill="off"
          onChange={ handleInputChange }
        />

        <input
          name="password"
          value={ password }
          className="auth__input"
          type="password"
          placeholder="password"
          autoComplete="off"
          autofill="off"
          onChange={ handleInputChange }
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={ loading }
        >Login</button>

        <div className="auth__social-networks">

          <h4 className="auth__subtitle">Login with social networks</h4>

          <div
            className="google-btn"
            onClick={ handleGoogleLogin }
          >

            <div className="google-icon-wrapper">
                <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>

            <p className="btn-text">
              <b>Sign in with google</b>
            </p>

          </div>

          <Link className="auth__link link" to="/auth/register">
            Create new account
          </Link>

        </div>

      </form>
    </>
  );
};

export default LoginScreen;
