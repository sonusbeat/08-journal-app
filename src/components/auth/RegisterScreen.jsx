import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";
import useForm from '../../hooks/useForm';
import { setError, unsetError } from '../../actions/ui';

const RegisterScreen = () => {
  const dispatch = useDispatch();

  const [ formValues, handleInputChange ] = useForm({
    name: "Daniel",
    email: "sonusbeat@gmail.com",
    password: "0123456789",
    password_confirmation: "0123456789",
  });

  const { name, email, password, password_confirmation } = formValues;

  const isFormValid = () => {
    if ( validator.isEmpty( name ) ) {
      dispatch( setError("Name should be present !") );
      return false;
    } else if ( name.trim().length <= 2) {
      dispatch( setError("Name should be greater than 2 characters !") );
      return false;
    } else if ( validator.isEmpty( email ) ) {
      dispatch( setError("Email should be present !") );
      return false;
    } else  if ( !validator.isEmail( email ) ) {
      dispatch( setError("Email format not valid !") );
      return false;
    } else  if ( validator.isEmpty( password ) ) {
      dispatch( setError("Password should be present !") );
      return false;
    } else  if ( password.trim().length < 8 ) {
      dispatch( setError("Password should be greater than 8 characters !") );
      return false;
    } else  if ( validator.isEmpty( password_confirmation ) ) {
      dispatch( setError("Password confirmation should be present !") );
      return false;
    } else  if ( password_confirmation.trim().length < 8 ) {
      dispatch( setError("Password confirmation should be greater than 8 characters !") );
      return false;
    } else  if ( password !== password_confirmation ) {
      dispatch( setError("Password and Password confirmation should be match each other !") );
      return false;
    }

    dispatch( unsetError() );

    return true;
  };

  const handleRegister = (e) => {
    e.preventDefault();

    if ( isFormValid() ) {
      console.log("Se envia el formulario");
    }
  };

  return (
    <>
      <div className="auth__alert-error mb-2">
        Mensaje de Error
      </div>

      <h3 className="auth__title">Register</h3>

      <form onSubmit={ handleRegister }>

        <input
          name="name"
          value={ name }
          className="auth__input"
          type="text"
          placeholder="name"
          autoComplete="off"
          onChange={ handleInputChange }
        />

        <input
          name="email"
          value={ email }
          className="auth__input"
          type="text"
          placeholder="email"
          autoComplete="off"
          onChange={ handleInputChange }
        />

        <input
          type="password"
          value={ password }
          className="auth__input"
          placeholder="password"
          name="password"
          autoComplete="off"
          onChange={ handleInputChange }
        />

        <input
          name="password_confirmation"
          value={ password_confirmation }
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          autoComplete="off"
          onChange={ handleInputChange }
        />

        <button
          className="btn btn-primary btn-block"
        >Register</button>

          <Link className="auth__link link" to="/auth/login">
            Already Registered ?
          </Link>
      </form>
    </>
  );
}

export default RegisterScreen;