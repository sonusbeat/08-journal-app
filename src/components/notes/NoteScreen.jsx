import { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
// import landscape from "../../images/landscape.jpg";

import NotesAppBar from './NotesAppBar';
import useForm from '../../hooks/useForm';
import { activeNote } from '../../actions/notes';

const NoteScreen = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm( note );
  const { title, body } = formValues;

  const activeId = useRef( note.id );

  useEffect(() => {
    if( note.id !== activeId.current ) {
      reset( note )
      activeId.current = note.id;
    }
  }, [ note, reset ]);

  useEffect(() => {
    dispatch( activeNote( formValues.id, { ...formValues } ) );
  }, [ dispatch, formValues ])

  return (
    <div className="notes">
      <div className="notes__main">
        <NotesAppBar />

        <div className="notes__content">

          <input
            className="notes__title-input"
            placeholder="Some awesome title"
            type="text"
            name="title"
            autoComplete="off"
            value={ title }
            onChange={ handleInputChange }
          />

          <textarea
            name="body"
            placeholder="What happend today ?"
            className="notes__textarea"
            value={ body }
            onChange={ handleInputChange }
          ></textarea>

          {
            ( note.url ) &&
            (
              <div className="notes__image">
                <img src={ note.url } alt="Landscape" />
              </div>
            )

          }
        </div>
      </div>
    </div>
  );
}

export default NoteScreen;