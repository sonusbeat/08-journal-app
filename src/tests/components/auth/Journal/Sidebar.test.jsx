import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import Sidebar from "../../../../components/journal/Sidebar";
import { startLogout } from "../../../../actions/auth";
import { startNewNote } from "../../../../actions/notes";

jest.mock("../../../../actions/auth", () => ({
  startLogout: jest.fn(),
}));

jest.mock("../../../../actions/notes", () => ({
  startNewNote: jest.fn(),
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

// Estado inicial del State de Redux
const initialState = {
  auth: {
    uid: "123458",
    name: "Daniel",
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    notes: [],
    active: null
  }
};

let store = mockStore( initialState );

// Simular el dispatch para las pruebas
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={ store }>
    <Sidebar />
  </Provider>
);

describe("Pruebas en <Sidebar />", () => {

  test("Deberia de coincidir con el Snapshot", () => {

    expect( wrapper ).toMatchSnapshot();

  });


  test("Deberia de llamar la acción startLogout", () => {

    // 1. Llamar la acción logout presionando el boton correspondiente
    wrapper.find("#logoutBtn").prop("onClick")();

    // 2. Se espera que la acción startLogout halle sido llamada
    expect( startLogout ).toHaveBeenCalled();

  });

  test("Deberia de llamar el startNewNote", () => {

    // 1. Llamar la acción logout presionando el boton correspondiente
    wrapper.find("#newEntryBtn").prop("onClick")();

    // 2. Se espera que la acción startNewNote halle sido llamada
    expect( startNewNote ).toHaveBeenCalledWith();

  });

});
