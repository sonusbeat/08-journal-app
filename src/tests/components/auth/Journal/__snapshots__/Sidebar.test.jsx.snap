// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar /> Deberia de coincidir con el Snapshot 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h2
      className="journal__sidebar-title"
    >
      <i
        className="far fa-user mr-1"
      />
      <span>
        Daniel
      </span>
    </h2>
    <button
      className="btn btn-primary"
      id="logoutBtn"
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </div>
  <div
    className="journal__sidebar-new-entry"
    id="newEntryBtn"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x mb-2"
    />
    <p>
      New Entry
    </p>
  </div>
  <div
    className="journal__sidebar-entries"
  />
</aside>
`;
